### PR TITLE
PROF-1037: Capture thread CPU time elapsed in scope

### PR DESCRIPTION
**Pull request by jbachorik**
Add thread CPU time elapsed while the scope was active to the Scope event.

Thread CPU time is comparatively cheap to obtain (~300ns per inocation) so we should be able to enhance the Scope event with the amount of CPU time spent in each recorded scope.

### CODE REVIEW

File: dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ScopeEvent.java [hunk 1]
```diff
@@ -46,12 +53,16 @@
...
     if (isEnabled()) {
       begin();
+      startCpuTime = ThreadCpuTime.get();
     }
   }
 
   @Override
   public void finish() {
     end();
+    if (startCpuTime > 0) {
```
Comment (mar-kolya) [inline dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ScopeEvent.java:63]: nit this should either be before end, or startCpuTime should be before begin

Comment (jbachorik) [reply dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ScopeEvent.java:63]: :+1:

File: dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ScopeEvent.java [hunk 2]
```diff
@@ -38,13 +39,18 @@
   @Label("Operation Name")
   private String operationName;
 
+  @Label("Thread CPU Time")
+  @Timespan
+  private volatile long cpuTime = 0L;
```
Comment (mar-kolya) [inline dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ScopeEvent.java:44]: My impression is that scopes are locked to given thread - is there any point in this being volatile?

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/Agent.java [hunk 3]
```diff
@@ -262,6 +265,36 @@ private static synchronized void installDatadogTracer() {
...
+      final ThreadMXBean mxBean = ManagementFactory.getThreadMXBean();
+      initializeMethod.setAccessible(true);
+      initializeMethod.invoke(
+          null,
+          new Callable<Long>() {
+            @Override
+            public Long call() throws Exception {
+              return mxBean.getCurrentThreadCpuTime();
+            }
+          });
```
Comment (mar-kolya) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/Agent.java:286]: Please move this code into separate package/class like we have for jmx, profiling and the agent - or maybe reuse something existing.

In fact maybe this whole thing can go into `initialize` method.
This class is already way overloaded and no need for odd jmx interactions here

File: dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ScopeEvent.java [hunk 4]
```diff
@@ -53,6 +59,9 @@ public void start() {
   public void finish() {
     end();
     if (shouldCommit()) {
+      if (cpuTime > 0) {
+        cpuTime = ThreadCpuTime.get() - cpuTime;
+      }
```
Comment (mar-kolya) [inline dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ScopeEvent.java:64]: Please add tests for this, especially considering that we already have `ScopeEventTest`

File: dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ThreadCpuTime.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  }
+
+  static long get() {
+    try {
+      return CPU_TIME_PROVIDER.call();
+    } catch (Exception ignored) {
+    }
+    return Long.MIN_VALUE;
+  }
+}
```
Comment (mar-kolya) [inline dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ThreadCpuTime.java:27]: This class should be unit-tested

File: dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ThreadCpuTime.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
+package datadog.opentracing.jfr.openjdk;
+
+import java.util.concurrent.Callable;
+
+public class ThreadCpuTime {
+  private static volatile Callable<Long> CPU_TIME_PROVIDER =
```
Comment (mar-kolya) [inline dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ThreadCpuTime.java:6]: nit: I think convention is to use `ALL_CAPS` only for constants.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/Agent.java [hunk 7]
```diff
@@ -262,6 +265,36 @@ private static synchronized void installDatadogTracer() {
...
+      log.debug("Thread CPU time provider enabled");
+    } catch (final Throwable ex) {
```
Comment (mar-kolya) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/Agent.java:288]: The whole `datadog.trace.agent.ot.jfr.openjdk` is compiled for java8 - so this will throw some form of `ClassFormatError` on java7 - please consider catching it separately like we do for profiling startup to provide more specific error message.

Alternatively you could move whole `datadog.trace.agent.ot.jfr.openjdk.ThreadCpuTime` into java7 compatible module since it in of itself doesn't need java8.

File: dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ScopeEvent.java (hunk 2 above)
Comment (jbachorik) [reply dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ScopeEvent.java:44]: If it is guaranteed that the event is created and committed from the same thread then this does not need to be volatile. I just wasn't sure there was such a guarantee.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/Agent.java (hunk 7 above)
Comment (jbachorik) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/Agent.java:288]: Let me think about the best place to put this. I was trying to find a good place without introducing new dependencies (which usually result in more transitive dependencies being pulled in and end up with something breaking compatibility ...).

File: dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ThreadCpuTime.java (hunk 6 above)
Comment (jbachorik) [reply dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ThreadCpuTime.java:6]: 👆

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/Agent.java (hunk 3 above)
Comment (jbachorik) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/Agent.java:286]: We (me and you) were discussing the class resolution rules in JVM and came to the conclusion that JLS does not forbid eagerly resolving all classes from the runtime constant pool - which would include JMX classes in `ThreadCpuTime` in this case - although it is quite improbable and none of the widely used JVM implementations are doing that. 
And since such resolution might cause JMX initialization at a wrong moment I came up with all this 'trampoline-like' stuff to be able to inject JMX code only when we are sure it is safe to do so. So, in short, this code can not go to `ThreadCpuTime.initialize()`

Not sure what would be gained by putting this into a different package/class. The reflection code would be as complex as it is now - we would only add one level of indirection there.

**Comment by jbachorik** [comment]
@devinsba The check would be quite tricky without reworking the initialization logic in `Agent`/`ProfilingAgent`. 

In the current implementation we first fire up `ProfilingAgent` initialization and only then we run (possibly asynchronous) JMX etc. initialization blocks. 
The `ThreadCpuTime.initialize()` routine needs to run within JMX initialization block to make sure we are not triggering JMX initialization at wrong time. But we should not be attempting to touch JFR classes from `Agent` itself since all interaction with JFR should be done in `ProfilingAgent`. 
However, `ProfilingAgent` has no idea about JMX etc. initialization. It is kind of chicken&egg problem.

I would really like to avoid major rewrite of initialization protocol for `Agent` in this PR. If there is a simple way to achieve the JFR check which would not possibly screw up JMX initialization I am not seeing, please advice.

File: dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ThreadCpuTime.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+  // must use Callable here since initialization is invoked from Agent which needs to be Java 7
+  // compatible
+  static void initialize(Callable<Long> provider) {
+    CPU_TIME_PROVIDER = provider;
+  }
+
+  static long get() {
+    try {
+      return CPU_TIME_PROVIDER.call();
+    } catch (Exception ignored) {
```
Comment (mar-kolya) [inline dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ThreadCpuTime.java:23]: If we are not expecting something happening here all the time maybe we should consider logging it on debug level?

File: dd-smoke-tests/profiling-integration-tests/src/test/groovy/datadog/smoketest/ProfilingIntegrationContinuousProfilesTest.groovy [hunk 9]
```diff
@@ -101,6 +104,30 @@ class ProfilingIntegrationContinuousProfilesTest extends AbstractSmokeTest {
...
+    def scopeEventIterable = scopeEvents.iterator().next()
+    def attribute = findCpuTimeAttribute(scopeEventIterable.type)
+    attribute != null
+
+    def accessor = scopeEventIterable.type.getAccessor(attribute)
+    def hasCpuTime = false
+    scopeEventIterable.every {
+      scopeEvent ->
+      hasCpuTime = hasCpuTime || accessor.getMember(scopeEvent).toLong()
+    }
```
[the remark this answers is not in the document]
Comment (jbachorik) [reply dd-smoke-tests/profiling-integration-tests/src/test/groovy/datadog/smoketest/ProfilingIntegrationContinuousProfilesTest.groovy:117]: I though about that - but it is not that easy. `Long.MIN_VALUE` is **really low value**.
Could have separate counters for existing and non-existing CPU time and then have an arbitrary threshold for what number of events with CPU time is still ok.

File: dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ThreadCpuTime.java (hunk 8 above)
Comment (jbachorik) [reply dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ThreadCpuTime.java]: This is gone now ...

File: dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ThreadCpuTime.java (hunk 5 above)
Comment (jbachorik) [reply dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ThreadCpuTime.java]: Done

File: dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ScopeEvent.java (hunk 4 above)
Comment (jbachorik) [reply dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ScopeEvent.java:66]: Added

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/Agent.java (hunk 3 above)
Comment (jbachorik) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/Agent.java]: This is now solved slightly differently. It is ok for ThreadCpuTime to stay in Java 7 source set.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/Agent.java (hunk 7 above)
Comment (jbachorik) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/Agent.java:288]: I think I found the place. This is already outdated.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/Agent.java [hunk 10]
```diff
@@ -262,6 +263,33 @@ private static synchronized void installDatadogTracer() {
     }
   }
 
+  private static synchronized void startJmx(final URL bootstrapURL) {
+    startJmxFetch(bootstrapURL);
+    /*
+     * Initialize the thread cpu time provider only if profiling is enabled.
+     * ATTENTION! If ever is the thread cpu time provider used outside of profiler this check should be revisited.
+     */
+    if (Config.get().isProfilingEnabled()) {
```
Comment (mar-kolya) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/Agent.java:272]: We do not use `Config` here - and IIRC there are awkward implications for doing that. Could you please move this check into `datadog.trace.common.util.ThreadCpuTimeAccess#enableJmx`?

Comment (jbachorik) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/Agent.java:272]: Well, I thought the idea was to avoid loading that class at all :)

File: dd-smoke-tests/profiling-integration-tests/src/test/groovy/datadog/smoketest/ProfilingIntegrationContinuousProfilesTest.groovy [hunk 11]
```diff
@@ -101,6 +104,29 @@ class ProfilingIntegrationContinuousProfilesTest extends AbstractSmokeTest {
...
+
+  private static Attribute<?> findCpuTimeAttribute(IType<IItem> type) {
+    def attribute = null
+    type.accessorKeys.forEach {
+      k, v ->
+        if (k.identifier == "cpuTime") {
+          attribute = k
+        }
+    }
+    return attribute
```
Comment (mar-kolya) [inline dd-smoke-tests/profiling-integration-tests/src/test/groovy/datadog/smoketest/ProfilingIntegrationContinuousProfilesTest.groovy:129]: Would `def accessor = Attribute.attr("cpuTime", UnitLookup.TIMESPAN).getAccessor(scopeEventIterable.type)` work instead of this?

File: dd-smoke-tests/profiling-integration-tests/src/test/groovy/datadog/smoketest/ProfilingIntegrationContinuousProfilesTest.groovy (hunk 9 above)
[the remark this answers is not in the document]
Comment (mar-kolya) [reply dd-smoke-tests/profiling-integration-tests/src/test/groovy/datadog/smoketest/ProfilingIntegrationContinuousProfilesTest.groovy:118]: I do not think this test is working: when I replace `100_000_000L` with `100_000L` in app code it doesn't fail.

In fact it will not fail if there are no events without cpu time.
It is also somewhat confusing to read - I feel that finding max of all cpuTimes (there is an aggregation API, isn't there?) would make it a bit more readable

[the remark this answers is not in the document]
Comment (mar-kolya) [reply dd-smoke-tests/profiling-integration-tests/src/test/groovy/datadog/smoketest/ProfilingIntegrationContinuousProfilesTest.groovy:118]: > I do not think this test is working: when I replace 100_000_000L with 100_000L in app code it doesn't fail.

Nm this, wrong number of zeros.

the `max` thing still stands, however.

[the remark this answers is not in the document]
Comment (mar-kolya) [reply dd-smoke-tests/profiling-integration-tests/src/test/groovy/datadog/smoketest/ProfilingIntegrationContinuousProfilesTest.groovy:118]: You may also want to consider changing this to 10ms - to get more events.

[the remark this answers is not in the document]
Comment (mar-kolya) [reply dd-smoke-tests/profiling-integration-tests/src/test/groovy/datadog/smoketest/ProfilingIntegrationContinuousProfilesTest.groovy:118]: ```scopeEvents.getAggregate(Aggregators.max("datadog.Scope", Attribute.attr("cpuTime", "cpuTime", UnitLookup.TIMESPAN))).longValue() >= 10_000_000L``` - something like this should work

File: dd-trace-ot/src/main/java/datadog/trace/common/util/ThreadCpuTimeAccess.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+       * point in time.
+       * Using reflection should alleviate this problem - no class constant to resolve during class load. The JMX
+       * thread cpu time provider will be loaded at exact moment when the reflection code is executed. Then it is up
+       * to the caller to ensure that it is safe to use JMX.
+       */
+      cpuTimeProvider =
+          (ThreadCpuTimeProvider)
+              Class.forName("datadog.trace.common.util.JmxThreadCpuTimeProvider")
+                  .getField("INSTANCE")
+                  .get(null);
```
Comment (mar-kolya) [inline dd-trace-ot/src/main/java/datadog/trace/common/util/ThreadCpuTimeAccess.java:39]: Nit: you do not really need instance there, instead you can call constructor here

File: dd-trace-ot/src/main/java/datadog/trace/common/util/ThreadCpuTimeProvider.java [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+
+/**
+ * A pluggable thread CPU time provider used by {@linkplain ThreadCpuTimeAccess}. {@linkplain
+ * ThreadCpuTimeAccess} may not use JMX classes (even via transitive dependencies) due to potential
+ * race in j.u.l initialization. Therefore it uses an abstract {@linkplain ThreadCpuTimeProvider}
+ * type to hold the actual implementation which may be switched between the {@linkplain
+ * ThreadCpuTimeProvider#DEFAULT} and {@linkplain JmxThreadCpuTimeProvider} on-the-fly once JMX is
+ * safe to use.
+ */
+abstract class ThreadCpuTimeProvider {
```
Comment (mar-kolya) [inline dd-trace-ot/src/main/java/datadog/trace/common/util/ThreadCpuTimeProvider.java:11]: You are creating instance of an abstract class: I think it is not so abstract then... I think this abstract can be dropped.

File: dd-smoke-tests/profiling-integration-tests/src/test/groovy/datadog/smoketest/ProfilingIntegrationContinuousProfilesTest.groovy (hunk 11 above)
Comment (jbachorik) [reply dd-smoke-tests/profiling-integration-tests/src/test/groovy/datadog/smoketest/ProfilingIntegrationContinuousProfilesTest.groovy:129]: D'oh ... Yes, it works.

File: dd-trace-ot/src/main/java/datadog/trace/common/util/ThreadCpuTimeAccess.java (hunk 12 above)
Comment (jbachorik) [reply dd-trace-ot/src/main/java/datadog/trace/common/util/ThreadCpuTimeAccess.java:39]: I don't want to create multiple instances if `enableJmx()` is called more times.

File: dd-smoke-tests/profiling-integration-tests/src/test/groovy/datadog/smoketest/ProfilingIntegrationContinuousProfilesTest.groovy (hunk 9 above)
[the remark this answers is not in the document]
Comment (jbachorik) [reply dd-smoke-tests/profiling-integration-tests/src/test/groovy/datadog/smoketest/ProfilingIntegrationContinuousProfilesTest.groovy:118]: Ok, done it

File: dd-trace-ot/src/main/java/datadog/trace/common/util/ThreadCpuTimeAccess.java (hunk 12 above)
Comment (mar-kolya) [reply dd-trace-ot/src/main/java/datadog/trace/common/util/ThreadCpuTimeAccess.java:44]: 🤷

File: dd-smoke-tests/profiling-integration-tests/src/test/groovy/datadog/smoketest/ProfilingIntegrationContinuousProfilesTest.groovy (hunk 9 above)
[the remark this answers is not in the document]
Comment (jbachorik) [reply dd-smoke-tests/profiling-integration-tests/src/test/groovy/datadog/smoketest/ProfilingIntegrationContinuousProfilesTest.groovy:115]: The aggregation looks ok. But I would run also min and filter the event collection first to exclude the events without valid cpu time

File: dd-trace-ot/src/main/java/datadog/trace/common/util/ThreadCpuTimeAccess.java [hunk 14]
```diff
@@ -8,19 +9,23 @@
...
   public static void disableJmx() {
     log.debug("Disabling JMX thread CPU time provider");
-    cpuTimeProvider = ThreadCpuTimeProvider.DEFAULT;
+    cpuTimeProvider = ThreadCpuTimeProvider.NONE;
   }
 
   /** Enable JMX based thread CPU time */
   public static void enableJmx() {
+    if (!Config.get().isProfilingEnabled()) {
+      log.info("Will not enable thread CPU time access. Profiling is disabled.");
```
Comment (mar-kolya) [inline dd-trace-ot/src/main/java/datadog/trace/common/util/ThreadCpuTimeAccess.java:26]: Let's do this on debug level, please - this seems irrelevant for clients that have profiling disabled.

**Comment by mar-kolya** [comment]
LGTM assuming it builds

File: dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ScopeEvent.java [hunk 15]
```diff
@@ -53,6 +61,9 @@ public void start() {
   public void finish() {
     end();
     if (shouldCommit()) {
+      if (cpuTime > 0) {
+        cpuTime = ThreadCpuTimeAccess.getCurrentThreadCpuTime() - cpuTime;
```
Comment (tylerbenson) [inline dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ScopeEvent.java:65]: Any risk of `finish()` being called more than once?

Comment (jbachorik) [reply dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ScopeEvent.java:65]: It is called solely from `ContinuableScope.close()` so as long as that scope is not closed multiple times we are fine.

Comment (mar-kolya) [reply dd-trace-ot/jfr-openjdk/src/main/java/datadog/opentracing/jfr/openjdk/ScopeEvent.java:65]: I'd say that if this happens things will not blow up but data we produce would be off. But arguably this is the best we can do in this situation anyway.
